core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,653 @@
+    Preconditions.checkArgument(duration >= 0, "duration must be greater than or equal to 0");
+    Preconditions.checkNotNull(unit, "unit");
+    return new CancellableContext(this, unit.toNanos(duration));
+  }
+
+  /**
+   * Create a new context with the given key value set. The new context will cascade cancellation
+   * from its parent.
+   */
+  public <V> Context withValue(Key<V> k1, V v1) {
+    return new Context(this, new Object[][]{{k1, v1}});
+  }
+
+  /**
+   * Create a new context with the given key value set. The new context will cascade cancellation
+   * from its parent.
+   */
+  public <V1, V2> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2) {
+    return new Context(this, new Object[][]{{k1, v1}, {k2, v2}});
+  }
+
+  /**
+   * Create a new context with the given key value set. The new context will cascade cancellation
+   * from its parent.
+   */
+  public <V1, V2, V3> Context withValues(Key<V1> k1, V1 v1, Key<V2> k2, V2 v2, Key<V3> k3, V3 v3) {
+    return new Context(this, new Object[][]{{k1, v1}, {k2, v2}, {k3, v3}});
+  }
+
+  /**
+   * Create a new context which copies the values of this context but does not propagate it's
+   * cancellation and is it's own independent root for cancellation.
+   */
+  public CancellableContext fork() {
+    return new Context(this).withCancellation();
+  }
+
+  /**
+   * Attach this context to the thread and make it {@link #current}, the previously current context
+   * will be restored when detach is called. It is allowed to attach contexts where
+   * {@link #isCancelled()} is {@code true}.
+   */
+  public void attach() {
+    contextStack.get().addLast(this);
+  }
+
+  /**
+   * Returns {@code true} if this context is {@link #current}, {@code false} otherwise.
+   */
+  public boolean isCurrent() {
+    return current() == this;
+  }
+
+  /**
+   * Detach the current context from the thread and restore the context that was previously
+   * attached to the thread as the 'current' context.
+   *
+   * @throws java.lang.IllegalStateException if this context is not {@link #current()}.
+   */
+  public void detach() {
+    ArrayDeque<Context> stack = contextStack.get();
+    if (stack.isEmpty()) {
+      if (this == ROOT) {
+        throw new IllegalStateException("Cannot detach root");
+      } else {
+        throw new IllegalStateException("Cannot detach non-root context when root is current");
+      }
+    }
+    if (stack.peekLast() != this) {
+      throw new IllegalStateException("Cannot detach a context that is not current");
+    }
+    stack.removeLast();
+  }
+
+  /**
+   * Is this context cancelled.
+   */
+  public boolean isCancelled() {
+    if (parent == null || !cascadesCancellation) {
+      return false;
+    } else {
+      return parent.isCancelled();
+    }
+  }
+
+  /**
+   * Return the cause of the cancellation or {@code null} if context was cancelled without a cause.
+   */
+  @Nullable
+  public Throwable cause() {
```

Comment:
We should make it clear what the receivers of cancellation are expected to do with this `cause`. I would suggest we say that the exception has already been "logged or handled" and that the exposure here is for further debugging.

I'd be happy if we didn't have the `cause()`, but assuming we do have it I want to make sure it is clearly an optional value to pay attention to.